hercules/python_simulators/battery.py

```diff
@@ -427,31 +513,104 @@ def control(self, P_avail, P_signal):
+        c_hi4 = P_avail
 
         # Lower constraints [kW]
-        c_lo1 = (self.E_min - self.E) / self.dt  # energy
+        # c_lo1 = (self.E_min - self.E) / self.dt  # energy
+        c_lo1 = self.SS_input_function_inverse((self.E_min - self.x[0, 0]) / self.dt)
         c_lo2 = self.P_min  # power
         c_lo3 = self.R_min * self.dt + self.P_charge  # ramp rate
 
         # High constraint is the most restrictive of the high constraints
-        c_hi = np.min([c_hi1, c_hi2, c_hi3, P_avail])
+        c_hi = np.min([c_hi1, c_hi2, c_hi3, c_hi4])
+        c_hi = np.max([c_hi, 0])
 
         # Low constraint is the most restrictive of the low constraints
         c_lo = np.max([c_lo1, c_lo2, c_lo3])
-        # TODO: force low constraint to be no higher than lowest high constraint
+        c_lo = np.min([c_lo, 0])
 
+        # TODO: force low constraint to be no higher than lowest high constraint
         if (P_signal >= c_lo) & (P_signal <= c_hi):
-            self.P_charge = P_signal
-            self.P_reject = 0
+            P_charge = P_signal
+            P_reject = 0
         elif P_signal < c_lo:
-            self.P_charge = c_lo
-            self.P_reject = P_signal - self.P_charge
+            P_charge = c_lo
+            P_reject = P_signal - P_charge
         elif P_signal > c_hi:
-            self.P_charge = c_hi
-            self.P_reject = P_signal - self.P_charge
+            P_charge = c_hi
+            P_reject = P_signal - P_charge
+
+        self.P_charge = P_charge
+        self.P_reject = P_reject
 
-    []
+        return P_charge, P_reject
+
+    def build_SS(self):
+        self.A = np.array([[-1 / self.tau_self_discharge]])
+        # B is the function in
+        self.C = np.array([[1, 0]]).T
+        self.D = np.array([[0, 1]]).T
+
+    def SS_input_function(self, P_charge):
+
+        # P_in is the amount of power that actually gets stored in the state E
+        # P_charge is the amount of power given to the charging physics
+
+        if P_charge >= 0:
+            P_in = self.eta_charge * P_charge
+        else:
+            P_in = P_charge / self.eta_discharge
+        return P_in
+
+    def SS_input_function_inverse(self, P_in):
+        if P_in >= 0:
+            P_charge = P_in / self.eta_charge
+        else:
+            P_charge = P_in * self.eta_discharge
+        return P_charge
+
+    def step_SS(self, u):
+        # Advance the state-space loop
+        xd = self.A * self.x + self.SS_input_function(u)
+        y = self.C * self.x + self.D * u
+
+        self.x = self.integrate(self.x, xd)
+        self.y = y
+
+    def integrate(self, x, xd):
+        # better integration -> use the closed form step response solution?
+        return x + xd * self.dt  # Euler integration
+
+    def calc_usage(self):
+
+        # Count rainflow cycles
+        cycles = rainflow.count_cycles(self.E_store)
+        self.total_cycle_usage = np.sum([cycle[0] * cycle[1] for cycle in cycles]) / self.E_max

```

Comment:
Maybe we could add a line explaining what is happening here? It's not immediately clear to me (although, it's a bit easier to follow after looking at the `rainflow.count_cycles` documentation :) ).